server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java

```diff
@@ -131,16 +148,87 @@ private static boolean isUnassignedDueToNewInitialization(ShardRouting routing)
+                    if (tierPreference.size() > 0) {
+                        Optional<DiscoveryNode> discoveryNode = allocateBasedOnTiersAndFilters(state, indexMetadata);
+                        if (discoveryNode.isPresent() == false) {
+                            return TIER_ROLE_CLASH;
+                        } else {
+                            LOGGER.debug(
+                                "Tier Check Outcome for [{}]: Found a viable node for tier: {}",
+                                routing.shardId(),
+                                discoveryNode.get().getId()
+                            );
+                        }
+                    } else {
+                        LOGGER.debug(
+                            "Tier Check Outcome for [{}]: Unknown - No tier configured: [{}]] from setting: All settings: [{}]",
+                            routing.shardId(),
+                            tierPreference,
+                            indexMetadata.getSettings()
+                        );
+                    }
+                } else {
+                    LOGGER.debug(
+                        "Tier Check Outcome for [{}]: Unknown - status: {}",
+                        routing.shardId(),
+                        routing.unassignedInfo().getLastAllocationStatus()
+                    );
+                }
+                return UNKNOWN;
+            })
+            .orElse(UNKNOWN);
+    }
+
+    /**
+     * Determines if a shard can be allocated to any node based on JUST tier preference and node filtering rules.
+     */
+    private static Optional<DiscoveryNode> allocateBasedOnTiersAndFilters(ClusterState state, IndexMetadata indexMetadata) {
+        List<String> tierPreference = indexMetadata.getTierPreference();
+        DiscoveryNodeFilters requireFilters = indexMetadata.requireFilters();
+        DiscoveryNodeFilters includeFilters = indexMetadata.includeFilters();
+        DiscoveryNodeFilters excludeFilters = indexMetadata.excludeFilters();
+
+        return state.nodes().getDataNodes().values().stream()
+            .filter(node -> {
+                if (node.getRoles().contains(DiscoveryNodeRole.DATA_ROLE)) {
+                    return true;
+                } else {
+                    for (String tierName : tierPreference) {
+                        for (DiscoveryNodeRole role : node.getRoles()) {
+                            if (role.roleName().equals(tierName)) {
+                                return true;
+                            }
+                        }
+                    }
+                }
+                return false;
+            })
+            .filter(node -> {if (requireFilters != null) return requireFilters.match(node); else return true;})
+            .filter(node -> {if (includeFilters != null) return includeFilters.match(node); else return true;})
+            .filter(node -> {if (excludeFilters != null) return excludeFilters.match(node); else return true;})
+            .findAny();
```

Comment:
This looks like a substantial duplication of logic from the Allocation Explain API - Should we just be using that in this indicator? There are some challenges to obtaining everything we need to call it correctly. Indicators are not injected with components and the explain action is run on the master node. The master node issue is the bigger of the two.

Reply:
I think for some of the deciders (the "simple" ones :) ) it's fine to locally ouput the explain decision and process those decisions (others like the `DiskThresholdDecider` works with settings configured on the master node so we'd need to come up with a different strategy).

For detecting the clash between the "_tier_preference" and "require.data" configuration for routing we don't have to check here if the appropriate tiers are present in the cluster (iterating through nodes and such). We already perform this detection in https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/cluster/metadata/MetadataMigrateToDataTiersRoutingService.java#L563, which is why I was suggesting we could possibly execute a dry_run for this migration service and check if there is an overlap between the indices that need migrating and the indices we detected as unassigned, but a different manual inspection of settings would also work.

I'd argue we should simply output a diagnosis that'll have a resolution of "migrate to data tiers routing".

What do you think?

Reply:
Looking at the `MetadataMigrateToDataTiersRoutingService` logic for migrating indices, I had some thoughts:

1) The logic presupposes a set of require/include allocation filters. The default is `data` but the user can specify any name for their own that should be migrated. In the case of a custom name other than `data` this detection logic breaks down. We're probably fine in that case, maybe later on we can check filter values and make a guess on a filter to use?
2) The service assumes that the coldest tier is always the first configured. Is that guaranteed?
3) If an index is capable of being migrated, does that necessarily mean that is the reason it is not allocated? I understand that this could be one of two cases: indices are using the "old" default configured filters for tiers or indices are being configured by users with their own hot/cold filters. Are we primarily concerned with just the first case here?

Reply:
Thanks for looking into this Jimmy. You make some great points. 
On:
1. The `data` attribute will likely tackle most cases we see, so I think it's fine to carry on inspecting this one (I think breaching into a blanket solution for _any_ other attribute could overstep over `require._id` which we set when shrinking for e.g., or a user using attributes to create sub-tiers)
2. The service will treat the coldest tier as the correct tier and migrate the settings to the coldest configuration which is where the index should go.
3. We'll likely have to evaluate a chain of deciders output - the ones we evaluate prior to the data tier conflict inspector will rule out _some_ of the other causes. It might not always be the case that this is root cause (it's hard to tell by just looking at a decider) but fixing this will not hurt the cases where there's another underlying issue
4. I don't think I understand what's former and latter in your question? We're concerned by a **mixed** usage of node attributes and tier preference which leads to problems of indices not moving to the desired tier or in certain circumstances unassigned indices.

Hope this helps? Let me know if I missed something

Reply:
> The service will treat the coldest tier as the correct tier and migrate the settings to the coldest configuration which is where the index should go.

The thing I was wondering about specifically is that the list of tiers is parsed from the preference setting, and the tier name that comes first is assumed to be the coldest, even if it is not actually the coldest. Do we have any guarantees around ensuring that is the case? e.g. if the current logic sees a preference list of `data_hot, data_cold, data_warm` it looks like it would select `data_hot` as the "coldest" tier because it comes first

> I don't think I understand what's former and latter in your question? We're concerned by a mixed usage of node attributes and tier preference which leads to problems of indices not moving to the desired tier or in certain circumstances unassigned indices.

Sorry, I might have been confused about the old way tiering used to work. I couldn't remember if we had a prescribed method of doing tiering that we shipped with all clusters that needed migrating or if it has always been a user configured feature before first class support.

Reply:
>Do we have any guarantees around ensuring that is the case?

We don't validate this at the moment, just that the tiers in the list are valid tier names - maybe open an improvement ticket for this? 
I haven't seen any configuration like this so far so I think for now it's fine to move ahead with the assumption the first is the coldest.